File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    int[][] dungeons;
    boolean[] visited;
    int answer;
    
    public int solution(int k, int[][] dungeons) {
        this.dungeons = dungeons;
        this.visited = new boolean[dungeons.length];
        
        dfs(k, 0);
        
        return answer;
    }
    
    private void dfs(int tired, int count) {
        answer = Math.max(answer, count);
        
        for(int i = 0; i < dungeons.length; i++) {
            if(visited[i]) continue;
            
            if(tired < dungeons[i][0]) continue;
            
            visited[i] = true;
            
            dfs(tired - dungeons[i][1], count + 1);
            
            visited[i] = false;
        }
    }
 }

File: 프로그래머스/2/87946. 피로도/피로도.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int answer = 0;
bool visited[8];

void dfs(int k, int count, vector<vector<int>>& dungeons) {
    answer = max(answer, count);
    
    for(int i = 0; i < dungeons.size(); i++) {
        if(!visited[i] && dungeons[i][0] <= k) {
            visited[i] = true;
            dfs(k - dungeons[i][1], count + 1, dungeons);
            visited[i] = false;
        }
    }
}

int solution(int k, vector<vector<int>> dungeons) {
    
    dfs(k, 0, dungeons);
    
    return answer;
}